File: src/xianyu_cli/utils/display.py
"""Rich terminal display formatters for CLI output."""

from __future__ import annotations

from typing import Any

from rich.panel import Panel
from rich.table import Table
from rich.text import Text

from xianyu_cli.utils._common import console


def _credit_display(credit: int | str) -> str:
    """Convert credit level to a star display string."""
    try:
        level = int(credit) if credit else 0
    except (ValueError, TypeError):
        return str(credit) if credit else "-"
    if level <= 0:
        return "-"
    # Taobao credit tiers: 1-5 hearts, 6-10 diamonds, 11-15 crowns, 16-20 gold crowns
    if level <= 5:
        return "♡" * level
    if level <= 10:
        return "◆" * (level - 5)
    if level <= 15:
        return "♛" * (level - 10)
    return "★" * min(level - 15, 5)


def _seller_credit_cell(item: dict[str, Any]) -> str:
    """Build a compact credit info string for table display."""
    parts: list[str] = []
    credit = item.get("seller_credit", 0)
    if credit:
        parts.append(_credit_display(credit))
    good_rate = item.get("seller_good_rate", "")
    if good_rate:
        parts.append(f"好评{good_rate}")
    sold = item.get("seller_sold_count", 0)
    if sold:
        parts.append(f"已售{sold}")
    zhima = item.get("zhima_credit", "")
    if zhima:
        parts.append(zhima)
    return " ".join(parts) if parts else "-"


def print_items_table(items: list[dict[str, Any]], title: str = "搜索结果") -> None:
    """Print a table of items to stderr using Rich."""
    has_credit = any(item.get("seller_credit") for item in items)

    table = Table(title=title, show_lines=True)
    table.add_column("ID", style="dim", width=12)
    table.add_column("标题", style="bold", max_width=40)
    table.add_column("价格", style="green", justify="right")
    table.add_column("卖家", style="cyan")
    if has_credit:
        table.add_column("信用", style="magenta", max_width=30)
    table.add_column("位置", style="yellow")

    for item in items:
        row = [
            str(item.get("id", "")),
            item.get("title", ""),
            f"¥{item.get('price', '?')}",
            item.get("seller_name", ""),
        ]
        if has_credit:
            row.append(_seller_credit_cell(item))
        row.append(item.get("location", ""))
        table.add_row(*row)

    console.print(table)


def print_item_detail(item: dict[str, Any]) -> None:
    """Print a detailed item view to stderr using Rich."""
    lines: list[str] = []
    lines.append(f"[bold]{item.get('title', '未知商品')}[/bold]")
    lines.append(f"价格: [green]¥{item.get('price', '?')}[/green]")
    lines.append(f"卖家: [cyan]{item.get('seller_name', '')}[/cyan]")
    lines.append(f"位置: {item.get('location', '')}")
    lines.append(f"描述: {item.get('description', '')}")

    if item.get("images"):
        lines.append(f"图片: {len(item['images'])} 张")

    console.print(Panel("\n".join(lines), title="商品详情", border_style="blue"))


def print_conversations(conversations: list[dict[str, Any]]) -> None:
    """Print a list of message conversations."""
    table = Table(title="会话列表", show_lines=True)
    table.add_column("会话ID", style="dim", width=16)
    table.add_column("对方", style="cyan")
    table.add_column("最新消息", max_width=40)
    table.add_column("时间", style="yellow")

    for conv in conversations:
        table.add_row(
            str(conv.get("id", "")),
            conv.get("peer_name", ""),
            conv.get("last_message", ""),
            conv.get("time", ""),
        )

    console.print(table)


def print_orders_table(orders: list[dict[str, Any]]) -> None:
    """Print a table of orders."""
    table = Table(title="订单列表", show_lines=True)
    table.add_column("订单号", style="dim", width=16)
    table.add_column("商品", max_width=30)
    table.add_column("金额", style="green", justify="right")
    table.add_column("状态", style="yellow")
    table.add_column("角色")

    for order in orders:
        table.add_row(
            str(order.get("id", "")),
            order.get("title", ""),
            f"¥{order.get('amount', '?')}",
            order.get("status", ""),
            order.get("role", ""),
        )

    console.print(table)


def print_profile(profile: dict[str, Any]) -> None:
    """Print user profile info."""
    lines: list[str] = []
    lines.append(f"[bold]{profile.get('nickname', '未知用户')}[/bold]")
    lines.append(f"用户ID: [dim]{profile.get('user_id', '')}[/dim]")
    if profile.get("credit_score"):
        lines.append(f"芝麻信用: {profile['credit_score']}")
    if profile.get("item_count"):
        lines.append(f"在售商品: {profile['item_count']} 件")

    console.print(Panel("\n".join(lines), title="个人资料", border_style="cyan"))


def print_success(msg: str) -> None:
    """Print a success message."""
    console.print(f"[green]✓[/green] {msg}")


def print_error(msg: str) -> None:
    """Print an error message."""
    console.print(f"[red]✗[/red] {msg}")


def print_warning(msg: str) -> None:
    """Print a warning message."""
    console.print(f"[yellow]![/yellow] {msg}")

File: golang/utils/display.go
package utils

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"github.com/fatih/color"
	"github.com/mattn/go-runewidth"
)

var (
	Green     = color.New(color.FgGreen)
	Red       = color.New(color.FgRed)
	Yellow    = color.New(color.FgYellow)
	Cyan      = color.New(color.FgCyan)
	Dim       = color.New(color.Faint)
	Bold      = color.New(color.Bold)
	BoldGreen = color.New(color.Bold, color.FgGreen)
	BoldRed   = color.New(color.Bold, color.FgRed)
)

// PrintSuccess prints a success message to stderr.
func PrintSuccess(msg string) {
	fmt.Fprintf(os.Stderr, "%s %s\n", Green.Sprint("✓"), msg)
}

// PrintError prints an error message to stderr.
func PrintError(msg string) {
	fmt.Fprintf(os.Stderr, "%s %s\n", Red.Sprint("✗"), msg)
}

// PrintWarning prints a warning message to stderr.
func PrintWarning(msg string) {
	fmt.Fprintf(os.Stderr, "%s %s\n", Yellow.Sprint("!"), msg)
}

// PrintDim prints dimmed text to stderr.
func PrintDim(msg string) {
	fmt.Fprintln(os.Stderr, Dim.Sprint(msg))
}

// PrintCyan prints cyan text to stderr.
func PrintCyan(msg string) {
	fmt.Fprintln(os.Stderr, Cyan.Sprint(msg))
}

// CreditDisplay converts credit level to a star display string.
func CreditDisplay(credit interface{}) string {
	var level int
	switch v := credit.(type) {
	case int:
		level = v
	case float64:
		level = int(v)
	case string:
		if v == "" {
			return "-"
		}
		l, err := strconv.Atoi(v)
		if err != nil {
			return v
		}
		level = l
	default:
		return "-"
	}
	if level <= 0 {
		return "-"
	}
	if level <= 5 {
		return strings.Repeat("♡", level)
	}
	if level <= 10 {
		return strings.Repeat("◆", level-5)
	}
	if level <= 15 {
		return strings.Repeat("♛", level-10)
	}
	n := level - 15
	if n > 5 {
		n = 5
	}
	return strings.Repeat("★", n)
}

// ── CJK-aware table renderer ──

// padRight pads a string to the given display width, respecting CJK double-width chars.
func padRight(s string, width int) string {
	sw := runewidth.StringWidth(s)
	if sw >= width {
		return s
	}
	return s + strings.Repeat(" ", width-sw)
}

// truncateWidth truncates a string to fit within maxWidth display columns.
// Replaces newlines with spaces first.
func truncateWidth(s string, maxWidth int) string {
	// Collapse newlines and extra whitespace
	s = strings.ReplaceAll(s, "\n", " ")
	s = strings.ReplaceAll(s, "\r", "")

	sw := runewidth.StringWidth(s)
	if sw <= maxWidth {
		return s
	}
	return runewidth.Truncate(s, maxWidth-1, "…")
}

// table is a simple CJK-aware table renderer.
type table struct {
	headers []string
	rows    [][]string
	colMax  []int // max display width per column
}

func newTable(headers []string) *table {
	colMax := make([]int, len(headers))
	for i, h := range headers {
		colMax[i] = runewidth.StringWidth(h)
	}
	return &table{headers: headers, colMax: colMax}
}

func (t *table) addRow(cells []string) {
	row := make([]string, len(t.headers))
	for i := 0; i < len(t.headers) && i < len(cells); i++ {
		row[i] = cells[i]
		w := runewidth.StringWidth(cells[i])
		if w > t.colMax[i] {
			t.colMax[i] = w
		}
	}
	t.rows = append(t.rows, row)
}

func (t *table) render() {
	// Clamp max widths to reasonable limits
	for i := range t.colMax {
		if t.colMax[i] > 80 {
			t.colMax[i] = 80
		}
	}

	// Top border
	t.printBorder("┌", "┬", "┐")

	// Header row
	t.printRow(t.headers, true)

	// Header separator
	t.printBorder("├", "┼", "┤")

	// Data rows with separators
	for i, row := range t.rows {
		t.printRow(row, false)
		if i < len(t.rows)-1 {
			t.printBorder("├", "┼", "┤")
		}
	}

	// Bottom border
	t.printBorder("└", "┴", "┘")
}

func (t *table) printBorder(left, mid, right string) {
	parts := make([]string, len(t.colMax))
	for i, w := range t.colMax {
		parts[i] = strings.Repeat("─", w+2) // +2 for padding
	}
	fmt.Fprintf(os.Stderr, "%s%s%s\n", left, strings.Join(parts, mid), right)
}

func (t *table) printRow(cells []string, bold bool) {
	parts := make([]string, len(t.colMax))
	for i, w := range t.colMax {
		cell := ""
		if i < len(cells) {
			cell = cells[i]
		}
		// Truncate if wider than column
		cell = truncateWidth(cell, w)
		padded := padRight(cell, w)
		parts[i] = " " + padded + " "
	}
	line := "│" + strings.Join(parts, "│") + "│"
	if bold {
		fmt.Fprintln(os.Stderr, Bold.Sprint(line))
	} else {
		fmt.Fprintln(os.Stderr, line)
	}
}

// ── Public table printers ──

// PrintItemsTable prints a table of items to stderr.
func PrintItemsTable(items []map[string]interface{}, title string) {
	if title == "" {
		title = "搜索结果"
	}
	fmt.Fprintf(os.Stderr, "\n%s\n", Bold.Sprint(title))

	hasCredit := false
	for _, item := range items {
		if v, ok := item["seller_credit"]; ok && v != nil && v != "" && v != 0 {
			hasCredit = true
			break
		}
	}

	headers := []string{"ID", "标题", "价格", "卖家"}
	if hasCredit {
		headers = append(headers, "信用")
	}
	headers = append(headers, "位置")

	tbl := newTable(headers)
	for _, item := range items {
		row := []string{
			fmt.Sprintf("%v", item["id"]),
			truncateWidth(fmt.Sprintf("%v", item["title"]), 36),
			fmt.Sprintf("¥%v", item["price"]),
			truncateWidth(fmt.Sprintf("%v", item["seller_name"]), 16),
		}
		if hasCredit {
			row = append(row, CreditDisplay(item["seller_credit"]))
		}
		row = append(row, fmt.Sprintf("%v", item["location"]))
		tbl.addRow(row)
	}
	tbl.render()
}

// PrintItemDetail prints detailed item info to stderr.
func PrintItemDetail(item map[string]interface{}) {
	fmt.Fprintln(os.Stderr)
	fmt.Fprintf(os.Stderr, "╭── %s ──╮\n", Bold.Sprint("商品详情"))
	fmt.Fprintf(os.Stderr, "│ %s: %s\n", Bold.Sprint("标题"), item["title"])
	fmt.Fprintf(os.Stderr, "│ %s: %s\n", Bold.Sprint("价格"), Green.Sprintf("¥%v", item["price"]))
	fmt.Fprintf(os.Stderr, "│ %s: %s\n", Bold.Sprint("卖家"), Cyan.Sprintf("%v", item["seller_name"]))
	fmt.Fprintf(os.Stderr, "│ %s: %v\n", Bold.Sprint("位置"), item["location"])
	fmt.Fprintf(os.Stderr, "│ %s: %v\n", Bold.Sprint("描述"), item["description"])
	if images, ok := item["images"].([]string); ok && len(images) > 0 {
		fmt.Fprintf(os.Stderr, "│ %s: %d 张\n", Bold.Sprint("图片"), len(images))
	}
	fmt.Fprintln(os.Stderr, "╰────────────╯")
}

// PrintConversations prints conversation list to stderr.
func PrintConversations(convs []map[string]interface{}) {
	fmt.Fprintf(os.Stderr, "\n%s\n", Bold.Sprint("会话列表"))
	tbl := newTable([]string{"会话ID", "对方", "最新消息", "时间"})
	for _, conv := range convs {
		tbl.addRow([]string{
			fmt.Sprintf("%v", conv["id"]),
			fmt.Sprintf("%v", conv["peer_name"]),
			truncateWidth(fmt.Sprintf("%v", conv["last_message"]), 36),
			fmt.Sprintf("%v", conv["time"]),
		})
	}
	tbl.render()
}

// PrintOrdersTable prints an orders table to stderr.
func PrintOrdersTable(orders []map[string]interface{}) {
	fmt.Fprintf(os.Stderr, "\n%s\n", Bold.Sprint("订单列表"))
	tbl := newTable([]string{"订单号", "商品", "金额", "状态", "角色"})
	for _, o := range orders {
		tbl.addRow([]string{
			fmt.Sprintf("%v", o["id"]),
			truncateWidth(fmt.Sprintf("%v", o["title"]), 28),
			fmt.Sprintf("¥%v", o["amount"]),
			fmt.Sprintf("%v", o["status"]),
			fmt.Sprintf("%v", o["role"]),
		})
	}
	tbl.render()
}

// PrintProfile prints user profile to stderr.
func PrintProfile(profile map[string]interface{}) {
	fmt.Fprintln(os.Stderr)
	fmt.Fprintf(os.Stderr, "╭── %s ──╮\n", Bold.Sprint("个人资料"))
	fmt.Fprintf(os.Stderr, "│ %s\n", Bold.Sprintf("%v", profile["nickname"]))
	fmt.Fprintf(os.Stderr, "│ 用户ID: %s\n", Dim.Sprintf("%v", profile["user_id"]))
	if cs, ok := profile["credit_score"]; ok && cs != "" {
		fmt.Fprintf(os.Stderr, "│ 芝麻信用: %v\n", cs)
	}
	if ic, ok := profile["item_count"]; ok && ic != 0 {
		fmt.Fprintf(os.Stderr, "│ 在售商品: %v 件\n", ic)
	}
	fmt.Fprintln(os.Stderr, "╰────────────╯")
}

// ── Public table API for external packages ──

// ReplyTable is a public CJK-aware table for displaying seller replies.
type ReplyTable struct {
	inner *table
}

// NewReplyTable creates a new table with the given headers.
func NewReplyTable(headers []string) *ReplyTable {
	return &ReplyTable{inner: newTable(headers)}
}

// AddRow adds a row to the table.
func (rt *ReplyTable) AddRow(cells []string) {
	rt.inner.addRow(cells)
}

// Render prints the table to stderr.
func (rt *ReplyTable) Render() {
	rt.inner.render()
}

// TruncateDisplay truncates a string to fit within maxWidth display columns (CJK-aware).
func TruncateDisplay(s string, maxWidth int) string {
	return truncateWidth(s, maxWidth)
}
